pkg/restore.go

```diff
@@ -156,40 +174,167 @@ func (opt *vaultOptions) restoreVault(targetRef api_v1beta1.TargetRef) (*restic.
+
+	opt.restoreOptions.RestorePaths = []string{opt.interimDataDir}
+
+	resticWrapper, err := restic.NewResticWrapper(opt.setupOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	restoreOutput, err := resticWrapper.RunRestore(opt.restoreOptions, targetRef)
+	if err != nil {
+		return nil, err
+	}
+
+	// restore the vault snapshot from the interim directory running command like: vault operator raft snapshot restore backup.snap
+	// for different vault server using -force flag like: vault operator raft snapshot restore -force backup.snap
+	err = opt.restoreVaultSnapshot(session)
+	if err != nil {
+		return nil, err
+	}
+
+	// if -force is true, we're potentially dealing with restoring snapshot on a different vault server
+	// we must replace the current vault server's unseal keys & root token with the older ones that we saved during snapshot
+	if opt.force {
+		klog.Infoln("Potentially different VaultServer. Trying to migrate old unseal keys & root token.")
+		if err := opt.setVaultTokenKeys(vs); err != nil {
+			return nil, err
+		}
+		klog.Infoln("Successfully migrated old unseal keys & root token")
+	}
+
+	return restoreOutput, nil
+}
+
+func (opt *VaultOptions) restoreVaultSnapshot(session *sessionWrapper) error {
+	klog.Infoln("Trying to restore snapshot")
+	session.cmd.Args = append(session.cmd.Args, "operator", "raft", "snapshot", "restore")
+
+	// -force is required for different vault cluster snapshot restoration
+	if opt.force {
+		klog.Infoln("Potentially different VaultServer. Applying -force to restore snapshot")
+		session.cmd.Args = append(session.cmd.Args, "-force")
+	}
+
+	session.cmd.Args = append(session.cmd.Args, filepath.Join(opt.interimDataDir, VaultSnapshotFile))
+
+	session.sh.ShowCMD = false
+	session.setUserArgs(opt.vaultArgs)
+	session.sh.Command(VaultCMD, session.cmd.Args...)
+
+	if err := session.sh.Run(); err != nil {
+		return err
+	}
+
+	klog.Infoln("snapshot restored successfully")
+	return nil
+}
+
+func (opt *VaultOptions) setVaultTokenKeys(vs *vaultapi.VaultServer) error {
+	klog.Infoln("Trying to read, set unseal keys & root token")
+	keyPrefix, err := opt.getKeyPrefix()
+	if err != nil {
+		return err
+	}
+	opt.keyPrefix = keyPrefix
+
+	// create a new store interface
+	// for restore:
+	// i. read the unseal keys & root token from the interim directory
+	// ii. Set the unseal keys & root token to store based on the unseal mode
+	st, err := store.NewStore(opt.kubeClient, vs)
+	if err != nil {
+		return err
+	}
+
+	var keys []string
+	keys = append(keys, opt.tokenName())
+	for i := 0; i < int(vs.Spec.Unsealer.SecretShares); i++ {
+		keys = append(keys, opt.unsealKeyName(i))
+	}
+
+	for _, key := range keys {
+		value, err := opt.read(key)
+		if err != nil {
+			klog.Errorf("failed to read key %s with %s\n", key, err.Error())
+			return err
+		}
+
+		if err := st.Set(key, value); err != nil {
+			klog.Errorf("failed to set key %s with %s\n", key, err.Error())
```

Comment:
will it affect the backup process if a key is not set?

Reply:
yes, these are some valid points, let me update the PR!